cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AdapterHttpHandler.java

```diff
@@ -50,24 +53,56 @@
  * {@link co.cask.http.HttpHandler} for managing adapter lifecycle.
  */
 @Singleton
-@Path(Constants.Gateway.API_VERSION_3 + "/namespaces/{namespace-id}/adapters")
-public class AdapterLifecycleHttpHandler extends AbstractAppFabricHttpHandler {
-  private static final Logger LOG = LoggerFactory.getLogger(AdapterLifecycleHttpHandler.class);
+@Path(Constants.Gateway.API_VERSION_3 + "/namespaces/{namespace-id}")
+public class AdapterHttpHandler extends AbstractAppFabricHttpHandler {
+  private static final Logger LOG = LoggerFactory.getLogger(AdapterHttpHandler.class);
 
   private final AdapterService adapterService;
   private final NamespaceAdmin namespaceAdmin;
   @Inject
-  public AdapterLifecycleHttpHandler(Authenticator authenticator, AdapterService adapterService,
-                                     NamespaceAdmin namespaceAdmin) {
+  public AdapterHttpHandler(Authenticator authenticator, AdapterService adapterService,
+                            NamespaceAdmin namespaceAdmin) {
     super(authenticator);
     this.namespaceAdmin = namespaceAdmin;
     this.adapterService = adapterService;
   }
 
   /**
+   * Deploy a template. Allows users to update a template, but leaves upgrade specifics to the user.
+   * For example, if a program is renamed or new arguments are required, users must handle recreation of
+   * adapters themselves.
+   */
+  @PUT
+  @Path("/templates/{template-id}")
+  public void deployTemplate(HttpRequest request, HttpResponder responder,
+                             @PathParam("namespace-id") String namespaceId,
+                             @PathParam("template-id") String templateId) {
+    if (!namespaceAdmin.hasNamespace(Id.Namespace.from(namespaceId))) {
+      responder.sendString(HttpResponseStatus.NOT_FOUND,
+                           String.format("Namespace '%s' does not exist.", namespaceId));
+      return;
+    }
+    try {
+      adapterService.deployTemplate(Id.Namespace.from(namespaceId), templateId);
+      responder.sendString(HttpResponseStatus.OK, "Deploy Complete");
+    } catch (NotFoundException e) {
```

Comment:
We could just use the catches to set HttpResponseStatus and message and add finally clause to do the responder.sendString call to make sure method sending response properly.